### ISSUE TITLE
[RFC] Implement Except Set Operation

### ISSUE THREAD

**Issue by jtenner** [tomi/fromfrom#27]
A first-time contributor here! I would love to open up a pull request and implement some stuff! Looks like the `except` function might be the lowest hanging fruit, so I'd love to take a stab at it.

On the [lodash documentation](https://lodash.com/docs/4.17.11#without) page, the `without` function uses a signature like this:

```ts
const items = _.without(array, ...items);
```

How about something like this?

```ts
const squares = from([1, 2, 3])
  .except(2) // or maybe name it `without` instead?
  .map(e => e * e)
  .toArray();
```

**Comment by leeoniya** [comment]
probably would be more flexible as a function, and more uniform with the rest of the lib.

```js
.not(e => e == 2)
```

**Comment by jtenner** [comment]
I was just reading the `TODO.md` and agree with your thoughts. What sort of work can be done instead?

**Comment by jtenner** [comment]
Sidenote: some functions could just be shorthand aliases for things.

```ts
export const withoutIterable = <TItem>(
  iterable: Iterable<TItem>,
  items: TItem[],
): Iterable<TItem> => {
  return filterIterable(iterable, e => !items.includes(e));
}
```

Keeps the implementation minimal.

**Comment by tomi** [comment]
Hi and thank you for opening an issue. I'm happy to hear that you are willing to contribute!

What I had in mind with `except` is pretty much what you described, i.e. what lodash's `without` also does. But instead of expecting an array, we could accept any `Iterable<TItem>`. This way we can support other iterable objects besides `array`s. So the signature could be:

```ts
.except(items: Iterable<TItem>): Sequence<TItem> {...}
// Or .without, I'm okay with using either name
```

Also it might make sense to have an overloaded method that accepts a custom equality comparer to compare the items, but this can also be added later on. The signature for that would be:

```ts
.except(items: Iterable<TItem>, comparer: EqualityCompareFn<TItem>): Sequence<TItem> {...}
// where the type for compare function is:
type EqualityCompareFn<TItem> = (a: TItem, b: TItem) => boolean;
```

And regarding the implementation, I do like using the `filterIterable` 👍 The given items could be first converted to `Set` tho, as `Array.prototype.includes` is O(n) where as `Set.prototype.has` is O(log(n)) or O(1) (at least according to [this](https://stackoverflow.com/a/31092145)).

**Comment by jtenner** [comment]
Great! I'm sorry for taking a while to get back to this issue. Let me spin up a pull request and I'll make some changes.

Question, because my brain likes lots of clarification. How should I use an `EqualityCompareFn`? 

This is what I would personally expect when I use the `without` operation.

```ts
import { createFilterIterable } from "./filter";

export const createWithoutIterable = <TItem>(
  iterable: Iterable<TItem>,
  items: Set<TItem>,
): Iterable<TItem> => 
  createFilterIterable(iterable, item => !items.has(item));
```

Also, we would need to exhaust the underlying item set and determine the whole set for this to work.

**Comment by tomi** [comment]
No worries, there's no rush :)

The idea behind `EqualityCompareFn` would be to allow the user to provide a custom function that compares if two objects are equal, instead of using the [`sameValueZero`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Equality_comparisons_and_sameness#Same-value-zero_equality) comparison. However, `Set` does not support providing a custom function for comparing the equality of two items. So I think at this point we can forget that and add it later on if it's actually needed.

**Comment by jtenner** [comment]
#31 addresses this enhancement as far as I am capable of. Thanks for the quick responses.

**Comment by tomi** [comment]
Available in 1.2.0 🎉
